test/jasmine/tests/calcdata_test.js

```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
+                    type: 'category',
+                    categorymode: 'value descending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(15);
+                expect(gd.calcdata[0][1].y).toEqual(14);
+                expect(gd.calcdata[0][2].y).toEqual(12);
+                expect(gd.calcdata[0][3].y).toEqual(11);
+
+            });
+        });
+
+        describe('explicit category ordering', function() {
+
+            it('should output categories in explicitly supplied order, independent of trace order', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'array',
+                    categorylist: ['b','a','d','e','c']
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(13);
+                expect(gd.calcdata[0][1].y).toEqual(11);
+                expect(gd.calcdata[0][2].y).toEqual(14);
+                expect(gd.calcdata[0][3].y).toEqual(12);
+                expect(gd.calcdata[0][4].y).toEqual(15);
+            });
+
+            it('should output categories in explicitly supplied order, independent of trace order, pruned', function() {
+
+                Plotly.plot(gd, [{x: ['c',undefined,'e','b','d'], y: [15,11,12,null,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'array',
+                    categorylist: ['b','a','d','e','c']
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(13);
+                expect(gd.calcdata[0][1].y).toEqual(14);
+                expect(gd.calcdata[0][2].y).toEqual(15);
+            });
+
+            it('should output categories in explicitly supplied order even if not all categories are present', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'array',
+                    categorylist: ['y','b','x','a','d','z','e','c']
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(13);
+                expect(gd.calcdata[0][1].y).toEqual(11);
+                expect(gd.calcdata[0][2].y).toEqual(14);
+                expect(gd.calcdata[0][3].y).toEqual(12);
+                expect(gd.calcdata[0][4].y).toEqual(15);
+            });
+
+            it('should output categories in explicitly supplied order first, if not all categories are covered', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'array',
+                    categorylist: ['b','a','x','c']
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(13);
+                expect(gd.calcdata[0][1].y).toEqual(11);
+                expect(gd.calcdata[0][2].y).toEqual(15);
+
```

Comment:
Hmm. I think `gd.calcdata[0][2].y` should be `null` in this case and  `gd.calcdata[0][3].y` be equal to 15.